core/auth/src/main/java/software/amazon/awssdk/auth/signer/Aws4UnsignedPayloadSigner.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright 2010-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.auth.signer;
+
+import static software.amazon.awssdk.auth.signer.internal.SignerConstant.X_AMZ_CONTENT_SHA256;
+
+import software.amazon.awssdk.annotations.SdkPublicApi;
+import software.amazon.awssdk.auth.signer.params.Aws4SignerParams;
+import software.amazon.awssdk.core.interceptor.ExecutionAttributes;
+import software.amazon.awssdk.http.SdkHttpFullRequest;
+
+/**
+ * Exactly the same as {@link Aws4Signer} except if the request is being sent
+ * over HTTPS, then it returns the string <code>UNSIGNED-PAYLOAD</code> as the
+ * content SHA-256 so services that support it can avoid needing to calculate
+ * the value when authorizing the request.
+ * <p>
+ * Payloads are still signed for requests over HTTP to preserve the request
+ * integrity over a non-secure transport.
+ */
+@SdkPublicApi
+public final class Aws4UnsignedPayloadSigner extends Aws4Signer {
+
+    private Aws4UnsignedPayloadSigner() {
+        super();
+    }
+
+    public static Aws4UnsignedPayloadSigner create() {
```

Comment:
hmmm, I think we are gonna have the similar issue as before where it has two static `create()` methods: one from here and the other inherited from the parent. Does it have to extend `Aws4Signer`?

Reply:
Intellij is only showing one create() method. Can you point me to the old issue you are referring to?

Reply:
oh, both of them are no-arg, so we are good for now, but if we add a static create method with argument in the parent such as `create(String arg)`, this class will inherit it.

Reply:
I'd say we should avoid inheritance if not necessary.

Reply:
Right, customers won't expect polymorphic. let's say we added `Aws4Signer.create(String arg)` in the parent and when a customer trying to create a `Aws4UnsignedPayloadSigner`, there are going to be two static `create` methods popping up, 
- `Aws4UnsignedPayloadSigner.create()` (create `Aws4UnsignedPayloadSigner`)  
- `Aws4UnsignedPayloadSigner.create(String arg)` (create `Aws4Signer`). 

It's a bit confusing to customers which one to use.

I think static members are hidden in a sense that they are not overridable by child classes. If the static methods in both classes have the same signature then it will be hidden in the child class, on the other hand, if they have different signatures, the child class will "inherit" that method.

> To avoid inheritance, we need to use composition and override all public, protected methods in Aws4Signer.


Why do we need to override all the public methods in `Aws4Signer`? can we just create a `Aws4Signer` field in `Aws4UnsignedPayloadSigner`, let's say `delegate`, and call `delegate.sign()`?

Reply:
> Let's say we added Aws4Signer.create(String arg)

In this case, we would need to add `create(String arg)` in Aws4UnsignedPayloadSigner too as we would need to construct the base class object

> Why do we need to override all the public methods in Aws4Signer? 

Yep, we would delegate all the calls. I meant we need to define [methods like these](https://github.com/aws/aws-sdk-java-v2/blob/master/core/auth/src/main/java/software/amazon/awssdk/auth/signer/Aws4Signer.java#L85-L104) in Aws4UnsignedPayloadSigner and call delegate in the body. 
In future, we add `protected String calculateSignature(foo)` method to `Aws4Signer`, then we need to add the same method to `Aws4UnsignedPayloadSigner` too and call `delegate.calculateSignature(foo)`

Basically we have similar issues in both approaches. I see scenario 2 (adding method like calculateSignature) likely to happen more than scenario 1 (adding overloaded create method).

Reply:
Gotcha, I still think we should try to avoid inheritance as it can introduce more issues and is more fragile. 

- Even though we can kind of solve the awkward two static `create` issue by creating the same `create(String arg)` in this class,  but it's going be useless if this class doesn't need that argument for instantiation at all. -> I'd be fine if we know they are always going to have the same constructors.

- Inheritance by nature violates encapsulation. If we add more protected and public methods in `Aws4Signer` in the future, `Aws4UnsignedPayloadSigner` will inherit those methods no matter if they are needed in that class or not. -> I'd be fine if it's guaranteed that all inherited methods in `Aws4UnsignedPayloadSigner` make sense.

- While `Aws4Signer` is not designed to be extended, it can be extended by customers and that could cause unwanted results. I recall we have seen this kind of issue in the past. (We can probably mitigate it a little by making some of the methods `final`).

I think calling delegate and adding more methods in `Aws4UnsignedPayloadSigner` gives us more flexibility and less error prune with the little price of more code.

Reply:
> Inheritance by nature violates encapsulation. 

I agree with it but it makes sense in here as essentially `Aws4Signer` and `Aws4UnsignedPayloadSigner` are exactly same except that latter doesn't sign payload using `https`.

I am not fully convinced we should go with delegate approach here. Let's see what others think about this.

Reply:
Added a common base class for Aws4Signer and Aws4UnsignedPayloadSigner. Using composition requires more refactoring to signers as a `calculateContentHash` is used in [AbstractAws4Signer.dosign](https://github.com/aws/aws-sdk-java-v2/blob/810f0e14e078aa60b295a229b5109eed27a026c8/core/auth/src/main/java/software/amazon/awssdk/auth/signer/internal/AbstractAws4Signer.java#L74) method and just overriding calculateContentHash method in the impl classes won't work.